### ISSUE TITLE
Method $product->getUrlInStore() returning extremely long URLs, could be a bug

### ISSUE THREAD

**Issue by adampmoss** [magento/magento2#16273]
I feel like there may be a bug in the following file because the output of `$product->getUrlInStore()` is not what I'd expect. Query params related to the store are added to the URL in the following file:

https://github.com/magento/magento2/blob/9c14af7a97bcf338662de365d2613c7ca9dbf8c6/app/code/Magento/Store/Url/Plugin/RouteParamsResolver.php#L66

By setting an entire object (`$subject->getScope()`) as the `$storeCode`rather than just a string `default` it is causing a very messy looking URL.

### Preconditions
1. Have a Magento installation with more than 1 store
2. Add store code to URLs in configuration is set to "no"
3. Magento version 2.2.3

### Steps to reproduce
Use the following code where a product model is loaded and where you can see the result:

`<?= $product->getUrlInStore() ?>`

### Expected result
The URL produced for the product would look like:

```
https://www.domain.com/my-product.html?___store=default
```

### Actual result
The URL produced is actually very long winded:

```
https://www.domain.com/my-product.html?___store%5B_data%5D%5Bstore_id%5D=1&___store%5B_data%5D%5Bcode%5D=default&___store%5B_data%5D%5Bwebsite_id%5D=1&___store%5B_data%5D%5Bgroup_id%5D=1&___store%5B_data%5D%5Bname%5D=UK+Mainland&___store%5B_data%5D%5Bsort_order%5D=0&___store%5B_data%5D%5Bis_active%5D=1&___store%5B_data%5D%5Bavailable_currency_codes%5D%5B0%5D=GBP&___store%5B_data%5D%5Bbase_currency%5D%5B_data%5D%5Bcurrency_code%5D=GBP&___store%5B_data%5D%5Bcurrent_currency%5D%5B_data%5D%5Bcurrency_code%5D=GBP
```

**Comment by vasilii-b** [comment]
This can be reproduced on Magento 2.3-develop as well.
And actually from steps to reproduce there is another bug: Having _Add store code to URLs_ set to NO, still shows you the URL like `https://www.domain.com/my-product.html?___store=default`.

**Comment by vasilii-b** [comment]
According to the file `/app/code/Magento/Store/Url/Plugin/RouteParamsResolver.php` and its content, this issue is reproducible on `Magento 2.2-develop` and `Magento 2.1-develop`.

**Comment by magento-engcom-team** [comment]
Hi @adampmoss. Thank you for your report. 
The issue has been fixed in magento/magento2#16468 by @vasilii-b in 2.2-develop branch
Related commit(s):
 - [1115b3f6e39f9b4406ada0a6dae7ac60ea333da3](https://github.com/magento/magento2/commit/1115b3f6e39f9b4406ada0a6dae7ac60ea333da3)
 - [03afbc73b3ab69f763b38dadd7ddcfb0b766e763](https://github.com/magento/magento2/commit/03afbc73b3ab69f763b38dadd7ddcfb0b766e763)
 - [7e3c3fc3b1da25b611ae3d80e0408dcd3a0d0282](https://github.com/magento/magento2/commit/7e3c3fc3b1da25b611ae3d80e0408dcd3a0d0282)
 - [66ee8273028fb3e0e118cdaad6e6163bf2477361](https://github.com/magento/magento2/commit/66ee8273028fb3e0e118cdaad6e6163bf2477361)
 - [846f60542af6606965485ca5a3b37ebd286ff218](https://github.com/magento/magento2/commit/846f60542af6606965485ca5a3b37ebd286ff218)
 - [65b735b2ea78aa5f6d7287a49f134a8d60b3973e](https://github.com/magento/magento2/commit/65b735b2ea78aa5f6d7287a49f134a8d60b3973e)

The fix will be available with the upcoming 2.2.6 release.

**Comment by sidolov** [comment]
Hi @adampmoss. Thank you for your report. 
The issue has been fixed in magento/magento2#17261 by @mage2pratik in 2.3-develop branch
Related commit(s):
 - [d1164548da45b9e2bb8ec22140061c1220748c23](https://github.com/magento/magento2/commit/d1164548da45b9e2bb8ec22140061c1220748c23)
 - [3a39a1674bf06d80d7df3969ecc78f84abd218c1](https://github.com/magento/magento2/commit/3a39a1674bf06d80d7df3969ecc78f84abd218c1)
 - [16ae409bbf5892b40b1a009eca76e2f1bbb73037](https://github.com/magento/magento2/commit/16ae409bbf5892b40b1a009eca76e2f1bbb73037)
 - [4dc6aba5daa877d5a39fcb8423c7f7a66875e8fd](https://github.com/magento/magento2/commit/4dc6aba5daa877d5a39fcb8423c7f7a66875e8fd)

The fix will be available with the upcoming 2.3.0 release.

**Comment by vasilii-b** [comment]
@magento-engcom-team , @sidolov hello!!!
there was already a PR created for 2.3-develop - https://github.com/magento/magento2/pull/16309, earlier than one already merged #17261. 😱 what is going on, guys ? Should we check first if issue/PR already created ? 🤦‍♂️ 
that is totally unfair...

**Comment by sidolov** [comment]
Hi @adampmoss. Thank you for your report. 
The issue has been fixed in magento/magento2#16310 by @vasilii-b in 2.1-develop branch
Related commit(s):
 - [2f8f469b36c7390a602b7f9c9713b03a64f5184b](https://github.com/magento/magento2/commit/2f8f469b36c7390a602b7f9c9713b03a64f5184b)
 - [ecfbe2450f844fea4dcc2c45aacc72accf5ec796](https://github.com/magento/magento2/commit/ecfbe2450f844fea4dcc2c45aacc72accf5ec796)
 - [7833f434f8e2eed29bb7c18fd3ee8f4185fbc1fa](https://github.com/magento/magento2/commit/7833f434f8e2eed29bb7c18fd3ee8f4185fbc1fa)
 - [7c33686a277fef1f37e9a315762126a6db83879b](https://github.com/magento/magento2/commit/7c33686a277fef1f37e9a315762126a6db83879b)
 - [726b6e9fbc7f314014298d4ece14550dc69c0ef6](https://github.com/magento/magento2/commit/726b6e9fbc7f314014298d4ece14550dc69c0ef6)

The fix will be available with the upcoming 2.1.16 release.

**Comment by arion-p** [comment]
The issue has been reintroduced in commit 22edbb904ae3856afbf871cc6e196edbd3bb679b. The commit is supposed to revert "Customer related values are NULL for guests" ccba42089f1ca465cb43fc60339671db5e9a504a but it actually reverts this fix (at least partly)
